lib/src/openai_api/lib/chat.py

```diff
@@ -68,3 +71,97 @@ async def exec_structured_chat(
     )
 
     return output_parser.parse(output.content.replace('\n', ''))
+
+
+
+async def exec_assistant(
+    message: dict[str, str], conf: Configuration
+) -> str:
+    client = AsyncOpenAI(
+        api_key=conf.api_key,
+        max_retries=conf.max_retries,
+        organization=conf.organization,
+    )
+    assistant = await client.beta.assistants.create(
+        model=message["model"],
+        instructions=message["instructions"],
+    )
+    return assistant.id
+
+
+
+async def exec_run(
+    messages: List[dict[str, str]], conf: Configuration
+) -> AsyncGenerator:
+    client = OpenAI(
+        api_key=conf.api_key,
+        max_retries=conf.max_retries,
+        organization=conf.organization
+    )
+    for message in messages:
+        assistant = client.beta.assistants.retrieve(
+            assistant_id=message["assistant_id"]
+        )
+        # if thread is NaN, create one
+        if isinstance(message["thread_id"], float):
+            run = client.beta.threads.create_and_run(
+                assistant_id=message["assistant_id"],
+                thread={
+                    "messages": [
+                        {
+                            "role": message["role"],
+                            "content": message["content"]
+                        }
+                    ]
+                },
+                model=assistant.model,
+                temperature=conf.temperature,
+                top_p=conf.top_p,
+                max_completion_tokens=conf.max_tokens
+            )
+        else:
+            # create a message, we do not need a handle for user messages
+            client.beta.threads.messages.create(
+                thread_id=message["thread_id"],
+                role=message["role"],
+                content=message["content"]
+            )
+            # create a run
+            run = client.beta.threads.runs.create(
+                thread_id=message["thread_id"],
+                assistant_id=message["assistant_id"],
+                model=assistant.model,
+                temperature=conf.temperature,
+                top_p=conf.top_p,
+                max_completion_tokens=conf.max_tokens,
+            )
+        run_id = run.id
+        thread_id = run.thread_id
+        # poll until completed
+        # no better way to do this until OpenAI updates the API
+        while run.status != "completed" and run.status != "failed":
```

Comment:
Let's have a counter of total retries to avoid hanging. Put it 500 seconds (5000 iterations).